### PR TITLE
sunxi/nezha: extend UART driver; document usage and mxstatus

### PR DESCRIPTION
**Pull request by orangecms**

### CODE REVIEW

**Comment by orangecms** [comment]
This is weird... looks like we got a clippy version that upsets our toolchain.

**Comment by orangecms** [comment]
Just found this English version of the manual, though missing the special CSRs as of now:
https://github.com/rjiejie/XuanTie-doc/blob/master/XuanTieCseriesinstructionmanual.pdf

**Review by rminnich — DISMISSED** [review]

File: src/drivers/uart/src/sunxi.rs [hunk 1]
```diff
@@ -18,19 +18,19 @@ const RETRY_COUNT: u32 = 100_000;
...
 pub const UART0_LC: u32 = UART0 + 0xc;
 pub const UART0_LS: u32 = UART0 + 0x14;
 
 // p897
-pub const UART1: u32 = 0x02500400;
-pub const UART2: u32 = 0x02500800;
-pub const UART3: u32 = 0x02500C00;
-pub const UART4: u32 = 0x02501000;
-pub const UART5: u32 = 0x02501400;
+pub const UART1: u32 = 0x0250_0400;
```
[the remark this answers is not in the document]
Comment (orangecms) [reply src/drivers/uart/src/sunxi.rs:29]: That sounds sensible to me. I haven't used `enum` in Rust yet, will have a look.

[the remark this answers is not in the document]
Comment (daym) [reply src/drivers/uart/src/sunxi.rs:29]: The official way is to have a Peripherals struct where you have all those different register blocks inside, and a take function to be able to take them *once*.

See also https://www.anyleaf.org/blog/writing-embedded-firmware-using-rust (search for "take").

The idea is that rust will move the ownership of the `Peripheral` member to whoever is using it right now.

See also https://docs.rust-embedded.org/book/peripherals/singletons.html#how-do-we-do-this-in-rust

**Review by rminnich — APPROVED** [review]

**Pull request merged**